crates/ruff/src/rules/ruff/typing.rs

```diff
@@ -0,0 +1,337 @@
+                let Some(new_target) =
+                    TypingTarget::try_from_expr(element, semantic, locator, minor_version)
+                else {
+                    return false;
+                };
+                new_target.contains_any(semantic, locator, minor_version)
+            }),
+            TypingTarget::Annotated(element) | TypingTarget::Optional(element) => {
+                let Some(new_target) =
+                    TypingTarget::try_from_expr(element, semantic, locator, minor_version)
+                else {
+                    return false;
+                };
+                new_target.contains_any(semantic, locator, minor_version)
+            }
+            TypingTarget::ForwardReference(expr) => {
+                let Some(new_target) =
+                    TypingTarget::try_from_expr(expr, semantic, locator, minor_version)
+                else {
+                    return false;
+                };
+                new_target.contains_any(semantic, locator, minor_version)
+            }
+        }
+    }
+}
+
+/// Check if the given annotation [`Expr`] explicitly allows `None`.
+///
+/// This function will return `None` if the annotation explicitly allows `None`
+/// otherwise it will return the annotation itself. If it's a `Annotated` type,
+/// then the inner type will be checked.
+///
+/// This function assumes that the annotation is a valid typing annotation expression.
+pub(crate) fn type_hint_explicitly_allows_none<'a>(
+    annotation: &'a Expr,
+    semantic: &SemanticModel,
+    locator: &Locator,
+    minor_version: u32,
+) -> Option<&'a Expr> {
+    let Some(target) = TypingTarget::try_from_expr(annotation, semantic, locator, minor_version)
+    else {
+        return Some(annotation);
+    };
+    match target {
+        // Short circuit on top level `None`, `Any` or `Optional`
+        TypingTarget::None | TypingTarget::Optional(_) | TypingTarget::Any => None,
+        // Top-level `Annotated` node should check for the inner type and
+        // return the inner type if it doesn't allow `None`. If `Annotated`
+        // is found nested inside another type, then the outer type should
+        // be returned.
+        TypingTarget::Annotated(expr) => {
+            type_hint_explicitly_allows_none(expr, semantic, locator, minor_version)
+        }
+        _ => {
+            if target.contains_none(semantic, locator, minor_version) {
+                None
+            } else {
+                Some(annotation)
+            }
+        }
+    }
+}
+
+/// Check if the given annotation [`Expr`] resolves to `Any`.
+///
+/// This function assumes that the annotation is a valid typing annotation expression.
+pub(crate) fn type_hint_resolves_to_any(
+    annotation: &Expr,
+    semantic: &SemanticModel,
+    locator: &Locator,
+    minor_version: u32,
+) -> bool {
+    let Some(target) = TypingTarget::try_from_expr(annotation, semantic, locator, minor_version)
```

Comment:
Just a stylistic thing, but consider doing this match on `Some` in the `match target` below? Like:

```rust
match TypingTarget::try_from_expr(annotation, semantic, locator, minor_version) {
  None => false,
  Some(TypingTarget::Any) => true,
  ...
}
```

But not blocking, up to your discretion of course.

Reply:
I like this, updated at other relevant places as well. Thanks!